clang/lib/CIR/CodeGen/CIRGenTypes.cpp

```diff
@@ -86,10 +86,80 @@ mlir::Type CIRGenTypes::convertFunctionTypeInternal(QualType qft) {
   return cir::FuncType::get(SmallVector<mlir::Type, 1>{}, cgm.VoidTy);
 }
 
+// This is CIR's version of CodeGenTypes::addRecordTypeName. It isn't shareable
+// because CIR has different uniquing requirements.
+std::string CIRGenTypes::getRecordTypeName(const clang::RecordDecl *recordDecl,
+                                           StringRef suffix) {
+  llvm::SmallString<256> typeName;
+  llvm::raw_svector_ostream outStream(typeName);
+
+  PrintingPolicy policy = recordDecl->getASTContext().getPrintingPolicy();
+  policy.SuppressInlineNamespace = false;
+
+  if (recordDecl->getIdentifier()) {
+    if (recordDecl->getDeclContext())
+      recordDecl->printQualifiedName(outStream, policy);
+    else
+      recordDecl->printName(outStream, policy);
+
+    // Ensure each template specialization has a unique name.
+    if (auto *templateSpecialization =
+            llvm::dyn_cast<ClassTemplateSpecializationDecl>(recordDecl)) {
```

Comment:
THIS block feels pretty error-prone.  I find myself wondering why we can't just reuse the code that specializations use to print the RecordType entirely (like for diagnostics), instead of all this.

Reply:
Why does this feel error prone? I'm not sure which code you're suggesting we could use instead.

Reply:
The template-args can print... awkwardly sometimes and not consistently depending on placeholder syntax and sugaring.  The code we already have to do such printing is already pretty onerous.   See `ConvertTypeToDiagnosticString` for what we typically use for diagnostics (which is a little more convoluted, but interesting all the same).

Clang `TypePrinter(LangOptions()).print` also does similar work and does a pretty solid job of it, and has some options to make it easier to not be weird.  So we might try that?